dmq-node/app/Main.hs

```diff
@@ -101,49 +115,75 @@ runDMQ commandLineConfig = do
-                       (decodeRemoteAddress (maxBound @NodeToNodeVersion)))
-                    dmqLimitsAndTimeouts
-                    defaultSigDecisionPolicy
-          dmqNtCApps =
-            let sigSize _ = 0 -- TODO
-                maxMsgs = 1000 -- TODO: make this dynamic?
-                mempoolReader = Mempool.getReader sigId sigSize (mempool nodeKernel)
-                mempoolWriter = Mempool.getWriter sigId (pure ())
-                                                        (\_ _ -> Right () :: Either Void ())
-                                                        (\_ -> True)
-                                                        (mempool nodeKernel)
-             in NtC.ntcApps tracer dmqConfig
-                            mempoolReader mempoolWriter maxMsgs
-                            (NtC.dmqCodecs encodeReject decodeReject)
-          dmqDiffusionArguments =
-            diffusionArguments (if handshakeTracer
-                                  then WithEventType "Handshake" >$< tracer
-                                  else nullTracer)
-                               (if localHandshakeTracer
-                                  then WithEventType "Handshake" >$< tracer
-                                  else nullTracer)
-          dmqDiffusionApplications =
-            diffusionApplications nodeKernel
-                                  dmqConfig
-                                  dmqDiffusionConfiguration
-                                  dmqLimitsAndTimeouts
-                                  dmqNtNApps
-                                  dmqNtCApps
-                                  (policy policyRng)
-
-      Diffusion.run dmqDiffusionArguments
-                    (dmqDiffusionTracers dmqConfig tracer)
-                    dmqDiffusionConfiguration
-                    dmqDiffusionApplications
+    Diffusion.withIOManager \iocp -> do
+      let localSnocket'      = localSnocket iocp
+          mkStakePoolMonitor = connectToCardanoNode tracer localSnocket' snocketPath
+
+      withNodeKernel @StandardCrypto
+                     tracer
+                     dmqConfig
+                     psRng
+                     mkStakePoolMonitor $ \nodeKernel -> do
+        dmqDiffusionConfiguration <-
+          mkDiffusionConfiguration dmqConfig nt nodeKernel.stakePools.ledgerBigPeersVar
+
+        let sigSize :: Sig StandardCrypto -> SizeInBytes
+            sigSize _ = 0 -- TODO
+            mempoolReader = Mempool.getReader sigId sigSize (mempool nodeKernel)
+            dmqNtNApps =
+              let ntnMempoolWriter = Mempool.writerAdapter $
+                    Mempool.getWriter sigId
+                                      (poolValidationCtx $ stakePools nodeKernel)
+                                      (validateSig (hashKey . VKey))
+                                      SigDuplicate
+                                      (mempool nodeKernel)
```

Comment:
The advantage of the previous API (without `ExceptT`) was clearer to review with the new API it's not obvious when the API is used how invalid signatures are handled.  For the `node-to-node` mini-protocol invalid signatures should throw an exception for `node-to-client` just refused from the mempool.  The previous mempool just returned `Either [(txid, RefusedReason)] [txid]` (or something similar) where the caller could decide what to do.  With the new API, this is unfortuantelly hidden from the call side, which makes the reading/reviewing much harder.